File: app/src/main/java/com/mju/csmoa/util/room/viewmodel/SearchHistoryViewModel.java
package com.mju.csmoa.util.room.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.mju.csmoa.util.room.entity.SearchHistory;
import com.mju.csmoa.util.room.repository.SearchHistoryRepository;

import java.util.List;

public class SearchHistoryViewModel extends AndroidViewModel {

    private final SearchHistoryRepository searchHistoryRepository;
    private final LiveData<List<SearchHistory>> searchHistoryList;

    public SearchHistoryViewModel(@NonNull Application application) {
        super(application);
        searchHistoryRepository = new SearchHistoryRepository(application);
        searchHistoryList = searchHistoryRepository.getSearchHistoryList();
    }

    public LiveData<List<SearchHistory>> getSearchHistoryList() {
        return searchHistoryList;
    }

    public void insert(SearchHistory searchHistory) {
        searchHistoryRepository.insert(searchHistory);
    }
}

File: app/src/main/java/com/mju/csmoa/util/room/viewmodel/SearchHistoryViewModel.kt
package com.mju.csmoa.util.room.viewmodel

import androidx.lifecycle.*
import com.mju.csmoa.util.room.entity.SearchHistory
import com.mju.csmoa.util.room.repository.SearchHistoryRepository
import kotlinx.coroutines.launch

class SearchHistoryViewModel(private val repository: SearchHistoryRepository) : ViewModel() {

    val allSearchHistories: LiveData<List<SearchHistory>> =
        repository.allSearchHistories.asLiveData()

    // Launching a new coroutine to insert the data in a non-blocking way
    fun insert(searchHistory: SearchHistory) = viewModelScope.launch {
        repository.insert(searchHistory)
    }
}

class SearchHistoryViewModelFactory(private val repository: SearchHistoryRepository) :
    ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(SearchHistoryViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return SearchHistoryViewModel(repository) as T
        }
        throw IllegalArgumentException("Unknown viewModel class")
    }

}
